### PR TITLE
Write metrics to result file during eval

### PR DESCRIPTION
**Pull request by abaek**
If a user specifies a `result` parameter in the eval part of their config, then it will write all metrics to that file.

To: @jq @yinlou

### CODE REVIEW

File: training/src/main/scala/com/airbnb/aerosolve/training/pipeline/GenericPipeline.scala [hunk 1]
```diff
@@ -139,6 +139,23 @@ object GenericPipeline {
     val metrics = evalCompute(sc, config, cfgKey, isTrainingFunc)
 
     metrics.foreach(x => log.info(if (x._1 contains "THRESHOLD") x._1 else x.toString))
+
+    // Write metrics to result
+    try {
```
Comment (jq) [inline training/src/main/scala/com/airbnb/aerosolve/training/pipeline/GenericPipeline.scala]: please wrap it into class.

File: training/src/main/scala/com/airbnb/aerosolve/training/pipeline/GenericPipeline.scala [hunk 2]
```diff
@@ -139,6 +139,23 @@ object GenericPipeline {
...
+      val result: String = confi.getConfig(cfgKey).getString("result")
```
Comment (jq) [inline training/src/main/scala/com/airbnb/aerosolve/training/pipeline/GenericPipeline.scala]: you will need the config of output result path, i.e. if it has output_result_path, then save result this way

File: training/src/main/scala/com/airbnb/aerosolve/training/pipeline/GenericPipeline.scala (hunk 1 above)
Comment (yinlou) [inline training/src/main/scala/com/airbnb/aerosolve/training/pipeline/GenericPipeline.scala]: agreed. Could be something like this:

def outputResults(records to write): Unit = {
  // log
  records.map(e => log.info(e.toString))
  // Write to hdfs
  Your logic
}

Comment (abaek) [inline training/src/main/scala/com/airbnb/aerosolve/training/pipeline/GenericPipeline.scala]: If I wrap it into a class, im not sure if i should also move the logging there

**Comment by abaek** [comment]
Build doesn't seem to pass, but could you verify that this is the right place to put the code. I can fix the syntax afterwards. @jq

File: training/src/main/scala/com/airbnb/aerosolve/training/pipeline/GenericPipeline.scala [hunk 3]
```diff
@@ -156,6 +156,7 @@ object GenericPipeline {
     val isRegression = Try(cfg.getBoolean("is_regression")).getOrElse(false)
     val isMulticlass = Try(cfg.getBoolean("is_multiclass")).getOrElse(false)
     val metric = cfg.getString("metric_to_maximize")
+    val resultsOutputPath = Try(cfg.getString("results_output_path"))..getOrElse("")
```
Comment (jq) [inline training/src/main/scala/com/airbnb/aerosolve/training/pipeline/GenericPipeline.scala]: ..

File: training/src/main/scala/com/airbnb/aerosolve/training/pipeline/GenericPipeline.scala [hunk 4]
```diff
@@ -447,7 +449,8 @@ object GenericPipeline {
       isRegression: Boolean,
       isMulticlass: Boolean,
       metric: String,
-      isTraining: Example => Boolean) : Array[(String, Double)] = {
+      isTraining: Example,
+      resultsOutputPath: String => Boolean) : Array[(String, Double)] = {
```
Comment (jq) [inline training/src/main/scala/com/airbnb/aerosolve/training/pipeline/GenericPipeline.scala]: Example => Boolean and String

File: training/src/main/scala/com/airbnb/aerosolve/training/Evaluation.scala [hunk 5]
```diff
@@ -87,6 +87,24 @@ object Evaluation {
...
+                                   resultOutputPath: String) : Array[(String, Double)] = {
+    val metrics = evaluateBinaryClassification(records, buckets, evalMetric)
+    ResultUtil.writeResults(metrics, resultOutputPath)
+    metrics
+  }
+
+  def evaluateBinaryClassification(records : List[EvaluationRecord],
+                                   buckets : Int,
+                                   evalMetric : String,
+                                   resultOutputPath: String) : Array[(String, Double)] = {
```
Comment (jq) [inline training/src/main/scala/com/airbnb/aerosolve/training/Evaluation.scala]: this two are same?

File: training/src/main/scala/com/airbnb/aerosolve/training/pipeline/ResultUtil.scala [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+
+import org.apache.hadoop.conf.Configuration
+import org.apache.hadoop.fs.FileSystem
+
+/*
+ * Writes to result file
+ */
+object ResultUtil {
+  def writeResults(metrics: Array[(String, Double)], resultOutputPath: String) = {
+    val fileSystem = FileSystem.get(new java.net.URI(result), new Configuration())
```
Comment (jq) [inline training/src/main/scala/com/airbnb/aerosolve/training/pipeline/ResultUtil.scala]: try reuse https://github.com/airbnb/aerosolve/blob/master/training/src/main/scala/com/airbnb/aerosolve/training/pipeline/PipelineUtil.scala#L39

**Comment by jq** [comment]
ok to put it in current places. You may talk to yizheng about the format of result.

File: training/src/main/scala/com/airbnb/aerosolve/training/Evaluation.scala (hunk 5 above)
Comment (abaek) [inline training/src/main/scala/com/airbnb/aerosolve/training/Evaluation.scala]: One of them is for List and the other is for RDD

File: training/src/main/scala/com/airbnb/aerosolve/training/pipeline/GenericPipeline.scala (hunk 4 above)
Comment (abaek) [inline training/src/main/scala/com/airbnb/aerosolve/training/pipeline/GenericPipeline.scala]: I'm not sure what you mean, I thought this line was correct

File: training/src/main/scala/com/airbnb/aerosolve/training/pipeline/ResultUtil.scala (hunk 6 above)
Comment (abaek) [inline training/src/main/scala/com/airbnb/aerosolve/training/pipeline/ResultUtil.scala]: `PipelineUtil.saveAndCommitAsTextFile` didn't have exactly what i needed, but I used `PipelineUtil.writeStringToFile`

File: training/src/main/scala/com/airbnb/aerosolve/training/pipeline/GenericPipeline.scala (hunk 4 above)
Comment (abaek) [inline training/src/main/scala/com/airbnb/aerosolve/training/pipeline/GenericPipeline.scala]: nvm, i  understand

**Comment by abaek** [comment]
Failed a test when I ran `gradle test`, but I don't think it was related to my changes

screen shot 2016-07-26 at 3 50 16 pm

**Comment by abaek** [comment]
Actually, working on changing the format, don't review yet please

File: training/src/main/scala/com/airbnb/aerosolve/training/pipeline/GenericPipeline.scala [hunk 7]
```diff
@@ -156,6 +156,7 @@ object GenericPipeline {
     val isRegression = Try(cfg.getBoolean("is_regression")).getOrElse(false)
     val isMulticlass = Try(cfg.getBoolean("is_multiclass")).getOrElse(false)
     val metric = cfg.getString("metric_to_maximize")
+    val resultsOutputPath = Try(cfg.getString("results_output_path")).getOrElse("")
```
Comment (jq) [inline training/src/main/scala/com/airbnb/aerosolve/training/pipeline/GenericPipeline.scala]: please update the demo with results_output_path, this is a good test for the code too.

File: training/src/main/scala/com/airbnb/aerosolve/training/pipeline/ResultUtil.scala [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+import com.airbnb.aerosolve.training._
+
+/*
+ * Writes to result file
+ */
+object ResultUtil {
+  def writeResults(metrics: Array[(String, Double)], resultOutputPath: String) = {
+    // Put into a one line json format
+    val elements = metrics.map{ x =>
+      val metricStr = if (x._1 contains "THRESHOLD") x._1 else x.toString
```
Comment (jq) [inline training/src/main/scala/com/airbnb/aerosolve/training/pipeline/ResultUtil.scala]: move THRESHOLD out, and add Unit test and comments for the output format.

**Comment by abaek** [comment]
I changed the result output format and updated this PR. Apparently you can't have default arguments and overload method names in scala, so I created new methods and resued a lot of code from the old methods.

I still have to work on adding a unit test, and testing this code with the demo.

File: training/src/main/scala/com/airbnb/aerosolve/training/pipeline/GenericPipeline.scala (hunk 7 above)
Comment (abaek) [inline training/src/main/scala/com/airbnb/aerosolve/training/pipeline/GenericPipeline.scala]: done

File: training/src/main/scala/com/airbnb/aerosolve/training/pipeline/ResultUtil.scala (hunk 8 above)
Comment (abaek) [inline training/src/main/scala/com/airbnb/aerosolve/training/pipeline/ResultUtil.scala]: I added comments for the output format, but I wasn't sure what to test in a unit test

File: training/src/main/scala/com/airbnb/aerosolve/training/pipeline/GenericPipeline.scala (hunk 7 above)
Comment (jq) [inline training/src/main/scala/com/airbnb/aerosolve/training/pipeline/GenericPipeline.scala]: seems not yet push? i.e. demo https://github.com/airbnb/aerosolve/tree/master/demo/income_prediction

File: training/src/main/scala/com/airbnb/aerosolve/training/Evaluation.scala [hunk 9]
```diff
@@ -87,6 +88,95 @@ object Evaluation {
       .toArray
   }
 
+  def evaluateBinaryClassification(records : RDD[EvaluationRecord],
+                                   buckets : Int,
+                                   evalMetric : String,
+                                   resultsOutputPath: String) : Array[(String, Double)] = {
+    var metrics  = mutable.Buffer[(String, Double)]()
```
Comment (jq) [inline training/src/main/scala/com/airbnb/aerosolve/training/Evaluation.scala]: lots of code are dup with the other evaluateBinaryClassification, can you share code between them?

File: training/src/main/scala/com/airbnb/aerosolve/training/Evaluation.scala [hunk 10]
```diff
@@ -197,6 +330,16 @@ object Evaluation {
     metrics.append(("!TRAIN_PRECISION", if (ttp > 0.0) ttp / (ttp + tfp) else 0.0))
     metrics.append(("!TRAIN_F1", 2 * ttp / (2 * ttp + tfn + tfp)))
 
+    if (results != null) {
```
Comment (jq) [inline training/src/main/scala/com/airbnb/aerosolve/training/Evaluation.scala]: maybe it is better to report it with metrics.append, i.e. metricsAppend(metrics, results, name, value)

File: training/src/main/scala/com/airbnb/aerosolve/training/Evaluation.scala (hunk 9 above)
Comment (abaek) [inline training/src/main/scala/com/airbnb/aerosolve/training/Evaluation.scala]: done, code is never repeated

File: training/src/main/scala/com/airbnb/aerosolve/training/pipeline/GenericPipeline.scala [hunk 11]
```diff
@@ -470,7 +473,7 @@ object GenericPipeline {
     } else if (isMulticlass) {
       Evaluation.evaluateMulticlassClassification(records)
     } else {
-      Evaluation.evaluateBinaryClassification(records, bins, metric)
+      Evaluation.evaluateBinaryClassificationWithResults(records, bins, metric, resultsOutputPath)
```
Comment (jq) [inline training/src/main/scala/com/airbnb/aerosolve/training/pipeline/GenericPipeline.scala]: if resultsOutputPath is null, appendMetricAndResult may crash..

Comment (abaek) [inline training/src/main/scala/com/airbnb/aerosolve/training/pipeline/GenericPipeline.scala]: I got rid of `appendMetricAndResult`. Now I am just relying on the metrics buffer and `resultsOutputPath` is only being used in one place here `Evaluation#L78`: 
```
    if (resultsOutputPath != null) {
      ResultUtil.writeResults(resultsOutputPath, metrics.toArray, holdThresholdPrecisionRecall.toArray, trainThresholdPrecisionRecall.toArray)
    }
```

Comment (jq) [inline training/src/main/scala/com/airbnb/aerosolve/training/pipeline/GenericPipeline.scala]: nice, please push the update

Comment (abaek) [inline training/src/main/scala/com/airbnb/aerosolve/training/pipeline/GenericPipeline.scala]: Woops done

File: training/src/main/scala/com/airbnb/aerosolve/training/pipeline/ResultUtil.scala [hunk 12]
```diff
@@ -0,0 +1,113 @@
...
+      0.313343,
+      0.897908,
+      1.190191,
+      1.482474
+    ]
+    }
+ *
+ */
+object ResultUtil {
+  def writeResults(resultsOutputPath: String,
```
Comment (jq) [inline training/src/main/scala/com/airbnb/aerosolve/training/pipeline/ResultUtil.scala]: suggest you add unit test for ResultUtil in this PR, and make ResultUtil extendable on next PR

**Comment by jq** [comment]
👍

File: training/src/main/scala/com/airbnb/aerosolve/training/Evaluation.scala [hunk 13]
```diff
@@ -20,119 +20,25 @@ object Evaluation {
...
-    thresholdsAppend(trainPR, holdPR, holdThresholdPrecisionRecall, trainThresholdPrecisionRecall)
-    val resultsArray = results.sortWith((a, b) => a._1 < b._1).toArray
-    ResultUtil.writeResults(resultsOutputPath, resultsArray, holdThresholdPrecisionRecall.toArray, trainThresholdPrecisionRecall.toArray)
-
-    metrics
-      .sortWith((a, b) => a._1 < b._1)
-      .toArray
+    // Convert RDD to list
+    val recordsList = List.fromArray(records.collect());
+    evaluateBinaryClassificationWithResults(recordsList, buckets, evalMetric, resultsOutputPath)
```
Comment (saurfang) [inline training/src/main/scala/com/airbnb/aerosolve/training/Evaluation.scala]: @abaek what was the motivation to call `collect` here? was it just to unify the codepath? I thought there were two evaluations that can compute locally when result is small but distributed via RDD when dataset is large.

Comment (abaek) [inline training/src/main/scala/com/airbnb/aerosolve/training/Evaluation.scala]: The motivation was to make all of the `evaluateBinaryClassification` share one code path and eliminate duplicate code. When I tried to refactor the code, I had a hard time sharing code between RDD's and List.

@jq mentioned that Evaluation jobs were small enough that collecting the RDD was ok.

Comment (saurfang) [inline training/src/main/scala/com/airbnb/aerosolve/training/Evaluation.scala]: I see. It makes sense but I'm having trouble collecting them to driver and driver is struggling with the evaluation. I'm going to try making everything follow the RDD codepath instead.

Comment (abaek) [inline training/src/main/scala/com/airbnb/aerosolve/training/Evaluation.scala]: Oh ok sorry about that! I think I only tested it with Lists.

**Pull request merged**